sapirbitan/assignment4/assignment4.cpp

```diff
@@ -0,0 +1,112 @@
+#include <iostream>
+#include <vector>
+using namespace std;
+//this class contains the row and the column of the cell
+class Cell {
+public:
+	int row;
+	int column;
+};
+
+bool goodRowColumn(int row, int column, int num_of_rows, int num_of_columns) {
+	return (row >= 0 && row < num_of_rows && column >= 0 && column < num_of_columns);
+}
+//finds all of the adjacent cells
+vector<Cell> getNeighbors(int current_row, int current_column, int num_of_rows, int num_of_columns) {
+	vector<Cell> neighbors;
+	Cell temp_cell;
+	//top cell
+	if (goodRowColumn(current_row -1, current_column, num_of_rows, num_of_columns)) {
+		temp_cell.row = current_row -1;
+		temp_cell.column = current_column;
+		neighbors.push_back(temp_cell);
+	}
+	//bottum cell
+	if (goodRowColumn(current_row +1, current_column, num_of_rows, num_of_columns)) {
+		temp_cell.row = current_row + 1;
+		temp_cell.column = current_column;
+		neighbors.push_back(temp_cell);
+	}
+	//left cell
+	if (goodRowColumn(current_row, current_column-1, num_of_rows, num_of_columns)) {
+		temp_cell.row = current_row;
+		temp_cell.column = current_column-1;
+		neighbors.push_back(temp_cell);
+	}
+	//right cell
+	if (goodRowColumn(current_row, current_column + 1, num_of_rows, num_of_columns)) {
+		temp_cell.row = current_row;
+		temp_cell.column = current_column + 1;
+		neighbors.push_back(temp_cell);
+	}
+	return neighbors;//returns the valid adjacent cells
+}
+//checks if the island is bigger than one cell on the map
+void checkArea(int row, int column, vector<vector<bool>> &map, vector<vector<bool>> &been_there) {
+	if (!map[row][column])
+		return;
+	been_there[row][column] = true;//the only places we want to mark as been_there are the ones who are land
+								  //in order to know it belongs to the same island
+	vector<Cell> neighbors = getNeighbors(row, column,map.size(),map[0].size());
```

Comment:
I'm not sure that `map[0].size()` is a safe thing to do.

In your code you don't check whether the map is perfectly rectangular. You should add a check that your map[0].size() is the same for all rows before you call this function, or write code that will be able to handle non-rectangular maps.